### ISSUE TITLE
Firewall rule source_ranges remain undeleted in GCP when replaced with source_tags in a fw rule resource configuration. 

### ISSUE THREAD

**Issue by averbuks** [hashicorp/terraform-provider-google#8076]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version
Terraform v0.13.5
+ provider registry.terraform.io/hashicorp/google v3.50.0

### Affected Resource(s)

google_compute_firewall

### Initial Terraform Configuration  

```tf
resource "google_compute_firewall" "default" {
  project = "averbukh-gke-onprem-test-bed"
  name = "test-fw-rule"
  network = "default"

  allow {
    protocol = "tcp"
    ports    = ["80", "8080"]
  }

  source_ranges = ["10.10.0.0/24", "10.20.0.0/24"]
}

```
### Initial  `terraform apply`
```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # google_compute_firewall.default will be created
  + resource "google_compute_firewall" "default" {
      + creation_timestamp = (known after apply)
      + destination_ranges = (known after apply)
      + direction          = (known after apply)
      + enable_logging     = (known after apply)
      + id                 = (known after apply)
      + name               = "test-fw-rule"
      + network            = "default"
      + priority           = 1000
      + project            = "averbukh-gke-onprem-test-bed"
      + self_link          = (known after apply)
      + source_ranges      = [
          + "10.10.0.0/24",
          + "10.20.0.0/24",
        ]

      + allow {
          + ports    = [
              + "80",
              + "8080",
            ]
          + protocol = "tcp"
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

google_compute_firewall.default: Creating...
google_compute_firewall.default: Still creating... [10s elapsed]
google_compute_firewall.default: Creation complete after 13s [id=projects/averbukh-gke-onprem-test-bed/global/firewalls/test-fw-rule]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

### Removing `source_ranges` and adding `source_tags` instead
```
resource "google_compute_firewall" "default" {
  project = "averbukh-gke-onprem-test-bed"
  name = "test-fw-rule"
  network = "default"

  allow {
    protocol = "tcp"
    ports    = ["80", "8080"]
  }
  source_tags = ["my-tag"]
}
```

### Apply changes `terraform apply`
```
google_compute_firewall.default: Refreshing state... [id=projects/averbukh-gke-onprem-test-bed/global/firewalls/test-fw-rule]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # google_compute_firewall.default will be updated in-place
  ~ resource "google_compute_firewall" "default" {
        creation_timestamp      = "2020-12-23T05:54:36.856-08:00"
        destination_ranges      = []
        direction               = "INGRESS"
        disabled                = false
        id                      = "projects/averbukh-gke-onprem-test-bed/global/firewalls/test-fw-rule"
        name                    = "test-fw-rule"
        network                 = "https://www.googleapis.com/compute/v1/projects/averbukh-gke-onprem-test-bed/global/networks/default"
        priority                = 1000
        project                 = "averbukh-gke-onprem-test-bed"
        self_link               = "https://www.googleapis.com/compute/v1/projects/averbukh-gke-onprem-test-bed/global/firewalls/test-fw-rule"
        source_ranges           = [
            "10.10.0.0/24",
            "10.20.0.0/24",
        ]
        source_service_accounts = []
      ~ source_tags             = [
          + "my-tag",
        ]
        target_service_accounts = []
        target_tags             = []

        allow {
            ports    = [
                "80",
                "8080",
            ]
            protocol = "tcp"
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

google_compute_firewall.default: Modifying... [id=projects/averbukh-gke-onprem-test-bed/global/firewalls/test-fw-rule]
google_compute_firewall.default: Still modifying... [id=projects/averbukh-gke-onprem-test-bed/global/firewalls/test-fw-rule, 10s elapsed]
google_compute_firewall.default: Modifications complete after 13s [id=projects/averbukh-gke-onprem-test-bed/global/firewalls/test-fw-rule]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

As you can see from the output plan - source_tags are added, but ranges are not changed and remain same. 

### Actual Behavior

After removing `source_ranges` and adding `source_tags` for a fw rule configuration - source ranges remain same (not deleted). So now fw rule allows both ranges and tags when user expects only tags being allowed after TF config changes.  It brings a quite *big security risk*. 

### Expected Behavior

After removing `source_ranges` and adding `source_tags` for a fw rule configuration - source ranges deleted and source_tags added. 

### Steps to Reproduce

1. Create an ingress fw rule with `source_ranges` specified and deploy it.
2. Delete `source_ranges` and add `source_tags`, apply changes.
3. Verify in Cloud Console that ranges remain undeleted and source tags are being added.

**Comment by venkykuberan** [comment]
Since it is `Computed`, the provider can receive input from the user or API. If the user doesn't specify a value in the config, the provider will use the API value which is `0.0.0.0/0` in this case.  If you explicitly set a value provider pass it on API. 
if you set the value to empty, then its removing the `source_range` entries and replace with `0.0.0.0/0` as its a mandatory attribute. Does it help ?

**Comment by averbuks** [comment]
> Since it is `Computed`, the provider can receive input from the user or API. If the user doesn't specify a value in the config, the provider will use the API value which is `0.0.0.0/0` in this case. If you explicitly set a value provider pass it on API.
> if you set the value to empty, then its removing the `source_range` entries and replace with `0.0.0.0/0` as its a mandatory attribute. Does it help ?

Hi @venkykuberan 
Maybe I'm missing something, but my understanding is following:
1. API value is 0.0.0.0/0 in case of source_tags are not provided, when they are provider - source ranges should be empty. In the current case they are provided after config change. 
2. Again speaking of API value 0/0, but when I remove source_ranges the value is not replaced to 0/0, but remains unchanged
see in the output after config change
```
        source_ranges           = [
            "10.10.0.0/24",
            "10.20.0.0/24",
        ]
```
So basically the issue is:
 - when I deploy a FW rule with source tags and without ranges from the firts place - ranges is empty (as expected)
 - when I deploy with source_ranges and then remove them and add source_tags, ranges are remain in the configuration, but should be gone.

**Comment by c2thorn** [comment]
Hey @averbuks, thanks for taking the time to point this out. `source_ranges` is designed to accept values returned by the API if it is not set in configuration. The value can exist/change in the backend and users do not have to manage it in Terraform. If this seems incorrect let us know, but this has been the behavior I believe since around the beginning of the implementation.

Within this behavior, configuration values have priority, but have to be explicitly stated. Intent to remove is shown with `source_ranges=[]` and intent to not manage is shown by removing the field from configuration.

The API allows for both `source_ranges` and `source_tags` to be set simultaneously, so I do not think we can add any custom logic to this behavior when `source_tags` is set. We can however, be more explicit about the behavior `source_ranges` in documentation, so users can be more aware that they need to set `source_ranges=[]` in this scenario. What do you think?

**Comment by averbuks** [comment]
Hi @c2thorn, thank you for the detailed explanation. 
In my way of thinking that behaviour is incorrect, I did a few tests and these are my assumptions:
1. source_tags - are managed with terraform, if no source_tags specified but some present - terraform will detect it and delete unmanaged tags,
2. source_sa - same thing as for source tags
3. So finally when source_ranges are specified but tags or SAs are omitted we consider them being setup to [], there are 2 cases:
   a. we had source_tags in the configuration but changed to source_ranges, terraform will delete the tags we had before.
   b. configurations based on source ranges, but someone went to the console and added source_tags to the rule, terraform will detect configuration drift and delete the manually added tags on apply. 
Both A and B works as expected and to my mind absolutely correct,
4. source_ranges are managed by terraform, but deleted from configuration and tags/SAs are present in configuration - terraform does not delete the ranges keeping current API value (as reported in the ticket)
5. we have source_tags or SAs in the configuration, source_ranges are omitted in the tf configuration. But someone goes to the console and adds ranges, terraform IS NOT going to detect configuration drift and will not delete manually added changes. 

So 4 and 5 seem to be incorrect behaviour as for me, a lot of users are relying on terraform when someone changes TF managed resource manually. Generally speaking FW rule in current state is not quite idempotent and the same TF configuration may end up in resource with different configuration (for example when source_ranges were specified and then deleted)
I believe there is a technical challenge and we end up having that behaviour, if that's the case it would be good to have at least a note in the documentation.

Thank you.

**Comment by c2thorn** [comment]
Rewriting `source_ranges` to not be `Computed` is likely a breaking change and evaluation is needed to determine why it was set to `Computed` in the first place. Setting as `persistent-bug` to be triaged

**Comment by rileykarson** [comment]
https://github.com/hashicorp/terraform-provider-google/pull/147
